nomos-services/blend/src/edge/mod.rs

```diff
@@ -87,118 +91,557 @@ where
+where
+    Backend: BlendBackend<NodeId, RuntimeServiceId> + Sync,
+    NodeId: Clone + Eq + Hash + Send + Sync + 'static,
+    RuntimeServiceId: Clone,
+{
+    // Read the initial membership, expecting it to be yielded immediately.
+    let SessionEvent::NewSession(membership) = session_stream
+        .next()
+        .await
+        .ok_or(Error::SessionStreamClosed)?
+    else {
+        return Err(Error::UnexpectedSessionEvent);
+    };
+
+    check_edge_condition(&membership, settings.minimum_network_size.get() as usize)?;
+
+    let mut message_handler = MessageHandler::<Backend, NodeId, RuntimeServiceId>::new(
+        settings,
+        membership,
+        overwatch_handle.clone(),
+    );
+
+    notify_ready();
+
+    loop {
+        tokio::select! {
+            Some(SessionEvent::NewSession(membership)) = session_stream.next() => {
+                message_handler = handle_new_session(membership, settings, overwatch_handle)?;
+            }
+            Some(message) = messages_to_blend.next() => {
+                message_handler.handle_messages_to_blend(message).await;
             }
         }
+    }
+}
 
+/// Check if the local node is an edge node in the membership,
+/// and if the membership size is at least the minimal network size.
+fn check_edge_condition<NodeId>(
+    membership: &Membership<NodeId>,
+    minimal_network_size: usize,
+) -> Result<(), Error>
+where
+    NodeId: Eq + Hash,
+{
+    if membership.size() < minimal_network_size {
+        Err(Error::NetworkIsTooSmall(membership.size()))
+    } else if membership.contains_local() {
+        Err(Error::LocalIsCoreNode)
+    } else {
         Ok(())
     }
 }
 
-/// Blend a new message received from another service.
-async fn handle_messages_to_blend<NodeId, Rng, Backend, RuntimeServiceId>(
-    message: Vec<u8>,
-    cryptographic_processor: &mut CryptographicProcessor<NodeId, Rng>,
-    backend: &Backend,
-) where
+#[derive(Debug, thiserror::Error)]
+enum Error {
+    #[error("Session stream closed")]
+    SessionStreamClosed,
+    #[error("Unexpected session event received")]
+    UnexpectedSessionEvent,
+    #[error("Network is too small: {0}")]
+    NetworkIsTooSmall(usize),
+    #[error("Local node is a core node")]
+    LocalIsCoreNode,
+}
+
+/// Handle a new session.
+///
+/// It creates a new [`MessageHandler`] if the membership satisfies all the edge
+/// node condition defined in [`check_edge_condition`].
+/// Otherwise, it returns [`Error`].
+fn handle_new_session<Backend, NodeId, RuntimeServiceId>(
+    membership: Membership<NodeId>,
+    settings: &Settings<Backend, NodeId, RuntimeServiceId>,
+    overwatch_handle: &OverwatchHandle<RuntimeServiceId>,
+) -> Result<MessageHandler<Backend, NodeId, RuntimeServiceId>, Error>
+where
+    Backend: BlendBackend<NodeId, RuntimeServiceId>,
+    NodeId: Clone + Eq + Hash + Send + 'static,
+    RuntimeServiceId: Clone,
+{
+    check_edge_condition(&membership, settings.minimum_network_size.get() as usize)?;
```

Comment:
Yes same here, I think it would be much better if `check_edge_condition` would be called `new_message_handler` and it would fail if the required condition is not met. Then, if we accidentally delete this line, we would get a compilation issue since we would not know how to create a `MessageHandler`, assuming its constructor is private and can only be called from inside the checking function, so moving the type in a different module and perhaps expose a method on the type itself that would check for the condition. What do you think?